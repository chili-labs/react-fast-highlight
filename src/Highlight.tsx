import * as React from 'react';
import PropTypes from 'prop-types';
import hljs from 'highlight.js';
import BareHighlight from './BareHighlight';

interface Props {
  children: string;
  className?: string;
  languages?: Array<string>;
  worker?: Worker;
}

const Highlight = (props: Props) => {
  const { children, ...rest } = props;

  return (
    <BareHighlight {...rest} highlightjs={hljs}>
      {children}
    </BareHighlight>
  );
};

Highlight.defaultProps = {
  className: '',
  languages: [],
  worker: null,
};

Highlight.propTypes = {
  children: PropTypes.string.isRequired,
  className: PropTypes.string,
  languages: PropTypes.arrayOf(PropTypes.string),
  worker: PropTypes.object, // eslint-disable-line react/forbid-prop-types
};

export default Highlight;
