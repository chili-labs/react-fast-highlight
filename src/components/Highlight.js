/* @flow */

import React, { Component } from 'react';
import shallowCompare from 'react/lib/shallowCompare';
import hljs from 'highlight.js';
import cx from 'classnames';

type Props = {
  children: string,
  className?: string,
  languages?: Array<string>,
};

type State = {
  highlightedCode: ?string,
  language: ?string,
};

export default class Highlight extends Component<void, Props, State> {

  state: State = {
    highlightedCode: null,
    language: null,
  };

  componentDidMount() {
    this._highlightCode();
  }

  shouldComponentUpdate(nextProps: Props, nextState: State): boolean {
    return shallowCompare(this, nextProps, nextState);
  }

  componentDidUpdate() {
    this._highlightCode();
  }

  // $FlowIssue - get/set properties not yet supported
  get initialCode(): string {
    return this.props.children;
  }

  // $FlowIssue - get/set properties not yet supported
  get highlightCallback(): () => HighlightResult {
    let callback;

    if (this.props.languages && this.props.languages.length === 1) {
      const language:string = this.props.languages[0];
      callback = (resolve: Function) =>
        resolve(hljs.highlight(language, this.initialCode));
    } else {
      callback = (resolve: Function) =>
        resolve(hljs.highlightAuto(this.initialCode, this.props.languages));
    }

    return callback;
  }

  _highlightCode(): void {
    const promise = new Promise(this.highlightCallback);

    promise
      .then(result => this.setState({ highlightedCode: result.value, language: result.language }))
      .catch(error => console.error(error));
  }

  render(): ?ReactElement {
    const code: ?string = this.state.highlightedCode;
    const classes = cx(this.props.className, 'hljs', this.state.language);

    if (code) {
      return (
        <pre>
          <code className={classes} dangerouslySetInnerHTML={{ __html: code }} />
        </pre>
      );
    } else {
      return <pre><code className={classes}>{this.initialCode}</code></pre>;
    }
  }
}
