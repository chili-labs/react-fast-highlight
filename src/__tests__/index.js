/* eslint-env jest */
import React from 'react';
import Highlight from '../index';

test('exports default component', () => {
  expect(<Highlight>test</Highlight>).toMatchReactSnapshot();
});
