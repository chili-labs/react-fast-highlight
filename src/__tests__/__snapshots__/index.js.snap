// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exports default component 1`] = `
<pre>
  <code
    className="hljs"
  >
    test
  </code>
</pre>
`;
